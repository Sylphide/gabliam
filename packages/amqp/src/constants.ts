export const LIST_QUEUE_BEAN = `LIST_QUEUE_BEAN`;

export const TYPE = {
  RabbitController: 'RabbitControllerType'
};

export const METADATA_KEY = {
  cunit: '_RabbitCuni',
  RabbitController: '_RabbitController',
  RabbitHandler: '_RabbitHandler'
};

export const ERRORS_MSGS = {
  DUPLICATED_CONTROLLER_DECORATOR: `Cannot apply @RabbitController decorator multiple times.`,
  DUPLICATED_CUNIT_DECORATOR: `Cannot apply @CUnit decorator multiple times.`
};

export const UNDEFINED_VALUE = '$$__##UNDEFINED##__$$';
