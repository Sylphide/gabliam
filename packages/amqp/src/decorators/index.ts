export * from './rabbitmq';
export * from './cunit';
