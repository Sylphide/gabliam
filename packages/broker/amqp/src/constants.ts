export const TYPE = {
  RabbitController: 'RabbitControllerType',
};

export const METADATA_KEY = {
  cunit: '_RabbitCuni',
  RabbitController: '_RabbitController',
  RabbitcontrollerParameter: '_Rabbitcontroller-parameter',
  RabbitHandler: '_RabbitHandler',
};

export const ERRORS_MSGS = {
  DUPLICATED_CONTROLLER_DECORATOR: `Cannot apply @RabbitController decorator multiple times.`,
  DUPLICATED_CUNIT_DECORATOR: `Cannot apply @CUnit decorator multiple times.`,
};

export const UNDEFINED_VALUE = '$$__##UNDEFINED##__$$';
