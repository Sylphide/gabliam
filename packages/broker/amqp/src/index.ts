export * from './amqp-connection';
export * from './amqp-manager';
export * from './errors';
export * from './interfaces';
export * from './metadatas';
export * from './queue';
export { AmqpPlugin as default };

import { AmqpPlugin } from './amqp-plugin';
