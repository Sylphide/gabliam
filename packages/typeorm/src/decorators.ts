import { register } from '@gabliam/core';
import {
  ChildEntity as typeormChildEntity,
  Entity as typeormEntity,
  EntityOptions,
} from 'typeorm';
import { METADATA_KEY, TYPE } from './constant';

/**
 * Set the connection of the entitie
 */
export function CUnit(name: string) {
  return (target: any) => {
    let metadata: string[] = [];
    if (Reflect.hasMetadata(METADATA_KEY.cunit, target) === true) {
      metadata = Reflect.getMetadata(METADATA_KEY.cunit, target);
    }
    metadata.push(name);

    Reflect.defineMetadata(METADATA_KEY.cunit, metadata, target);
    return target;
  };
}

/**
 * This decorator is used to mark classes that will be an entity (table or document depend on database type).
 * Database schema will be created for all classes decorated with it, and Repository can be retrieved and used for it.
 */
export function Entity(name?: string, options?: EntityOptions) {
  return (target: any) => {
    register(TYPE.Entity, { id: target, target })(target);
    typeormEntity(name, options)(target);
  };
}

/**
 * Special type of the table used in the single-table inherited tables.
 */
export function ChildEntity() {
  return (target: any) => {
    register(TYPE.Entity, { id: target, target })(target);
    typeormChildEntity()(target);
  };
}
