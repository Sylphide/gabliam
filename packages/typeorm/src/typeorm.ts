export * from 'typeorm';

import {
  Entity,
  ClassEntityChild,
  SingleEntityChild,
  ClosureEntity
} from './decorators';

export { Entity, ClassEntityChild, SingleEntityChild, ClosureEntity };
