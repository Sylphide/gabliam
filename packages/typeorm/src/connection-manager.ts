import { ValueExtractor } from '@gabliam/core';
import { createConnections } from 'typeorm';
import { METADATA_KEY } from './constant';
import { Connection, ConnectionOptions } from './typeorm';

export class ConnectionManager {
  private connections!: Connection[];

  constructor(
    private connectionOptions: ConnectionOptions[],
    private entities: Function[],
    private valueExtractor: ValueExtractor
  ) {}

  async open() {
    const connectionOptions = this.connectionOptions.map(c => {
      return {
        ...c,
        entities: Array.isArray(c.entities) ? c.entities : [],
        name: c.name ? c.name : 'default',
      };
    });

    // add entity to the correct connection
    for (const entity of this.entities) {
      const cunits: string[] = Reflect.getMetadata(
        METADATA_KEY.cunit,
        entity
      ) || ['default'];

      for (let cunit of cunits) {
        cunit = this.valueExtractor(cunit, cunit);

        let index = connectionOptions.findIndex(c => c.name === cunit);

        if (index === -1 && cunit === 'default') {
          index = 0;
        }

        if (index === -1) {
          throw new Error(`CUnit ${cunit} not found`);
        }
        (<any>connectionOptions)[index].entities.push(entity);
      }
    }

    this.connections = await createConnections(connectionOptions);
  }

  async close() {
    await Promise.all(this.connections.map(c => c.close()));
  }

  getConnection(name: string) {
    const connection = this.connections.find(c => c.name === name);
    if (!connection) {
      throw new Error(`Connection ${name} not found`);
    }
    return connection;
  }

  getDefaultConnection() {
    const connection = this.connections.find(c => c.name === 'default');
    if (connection === undefined) {
      return this.connections[0];
    } else {
      return connection;
    }
  }
}
