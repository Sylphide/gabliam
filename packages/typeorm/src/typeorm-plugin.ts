import {
  Scan,
  Registry,
  Plugin,
  GabliamPlugin,
  ValueRegistry,
  Container
} from '@gabliam/core';
import { TYPE, ENTITIES_TYPEORM } from './constant';
import { ConnectionManager } from './connection-manager';

@Plugin()
@Scan()
export class TypeOrmPlugin implements GabliamPlugin {
  bind(container: Container, registry: Registry) {
    const entities = registry
      .get<ValueRegistry>(TYPE.Entity)
      .map(({ target }) => {
        return target;
      });
    container.bind<any>(ENTITIES_TYPEORM).toConstantValue(entities);
  }

  async destroy(container: Container, registry: Registry) {
    const connection = container.get(ConnectionManager);
    await connection.close();
  }
}
