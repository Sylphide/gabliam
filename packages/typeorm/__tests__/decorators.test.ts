import { METADATA_KEY as CORE_METADATA_KEY } from '@gabliam/core/lib/constants';
import { RegistryMetada } from '@gabliam/core/lib/interfaces';
import { METADATA_KEY } from '../src/constant';
import { ChildEntity, CUnit, Entity } from '../src/decorators';

[Entity, ChildEntity].forEach(deco => {
  test(`${deco.name} decorators`, () => {
    @deco()
    class TestEntity {}

    const entityMetadata: RegistryMetada = Reflect.getMetadata(
      CORE_METADATA_KEY.register,
      TestEntity
    );

    expect(entityMetadata).toMatchSnapshot();
  });
});

describe('CUnit decorators', () => {
  test('should add CUnit metadata to a class when decorating a method with @CUnit', () => {
    @CUnit('default')
    class TestEntity {}

    const entityMetadata: RegistryMetada = Reflect.getMetadata(
      METADATA_KEY.cunit,
      TestEntity
    );

    expect(entityMetadata).toMatchSnapshot();
  });

  test('should add CUnit metadata to a class when decorating multiple times with @CUnit', () => {
    @CUnit('default')
    @CUnit('default2')
    class TestEntity {}

    const entityMetadata: RegistryMetada = Reflect.getMetadata(
      METADATA_KEY.cunit,
      TestEntity
    );

    expect(entityMetadata).toMatchSnapshot();
  });
});
