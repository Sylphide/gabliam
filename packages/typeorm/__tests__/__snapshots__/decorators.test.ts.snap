// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CUnit decorators should add CUnit metadata to a class when decorating a method with @CUnit 1`] = `
Array [
  "default",
]
`;

exports[`CUnit decorators should add CUnit metadata to a class when decorating multiple times with @CUnit 1`] = `
Array [
  "default2",
  "default",
]
`;

exports[`ChildEntity decorators 1`] = `
Object {
  "type": "EntityTypeOrm",
  "value": Object {
    "id": [Function],
    "target": [Function],
  },
}
`;

exports[`Entity decorators 1`] = `
Object {
  "type": "EntityTypeOrm",
  "value": Object {
    "id": [Function],
    "target": [Function],
  },
}
`;
